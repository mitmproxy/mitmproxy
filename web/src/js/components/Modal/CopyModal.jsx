import React, { Component } from "react"
import { connect } from "react-redux"
import * as modalAction from "../../ducks/ui/modal"
import Clipboard from 'react-clipboard.js';
import * as flowsActions from "../../ducks/flows"


class PureCopyModal extends Component {

    constructor(props, context) {
        super(props, context)
        this.state = { title: 'Curl', curl: '' }
    }

    componentDidMount() {
        const { flow, copyFlow } = this.props
        copyFlow(flow)
        .then(res => res.json())
        .then((result) => {
            this.setState({
                curl: result.curl
            })
          }
        )
      }

    render() {
        const { hideModal} = this.props
        const { title, curl } = this.state
        return (
            <div>
                <div className="modal-header">
                    <button type="button" className="close" data-dismiss="modal" onClick={() => {
                        hideModal()
                    }}>
                        <i className="fa fa-fw fa-times"></i>
                    </button>
                    <div className="modal-title">
                        <h4>{ title }</h4>
                    </div>
                </div>

                <div className="modal-body">
                    <div style={{overflowWrap: 'break-word'}} id="curl">{curl}</div>
                </div>

                <div className="modal-footer">
                    <Clipboard className="btn btn-primary" data-clipboard-target="#curl">copy</Clipboard>
                </div>
            </div>
        )
    }
}

export default connect(
    state => ({
        flow: state.flows.byId[state.flows.selected[0]],
    }),
    {
        hideModal: modalAction.hideModal,
        copyFlow: flowsActions.copy
    }
)(PureCopyModal)
