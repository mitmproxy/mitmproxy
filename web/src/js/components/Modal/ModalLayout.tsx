import * as React from "react"

type ModalLayoutProps = {
    children: React.ReactNode,
}

export default function ModalLayout ({ children}: ModalLayoutProps ) {
    return (
        <div>
        <div className="modal-backdrop fade in"></div>
        <div className="modal modal-visible" id="optionsModal" tabIndex={-1} role="dialog" aria-labelledby="options">
        <div className="modal-dialog modal-lg" role="document">
            <div className="modal-content">
                {children}
            </div>
        </div>
    </div>
        </div>
    )
}
