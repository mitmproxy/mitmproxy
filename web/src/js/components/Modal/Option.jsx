import React, { Component } from "react"
import PropTypes from "prop-types"
import { connect } from "react-redux"
import { update as updateOptions } from "../../ducks/options"
import { Key } from "../../utils"
import classnames from 'classnames'

const stopPropagation = e => {
    if (e.keyCode !== Key.ESC) {
        e.stopPropagation()
    }
}

BooleanOption.propTypes = {
    value: PropTypes.bool.isRequired,
    onChange: PropTypes.func.isRequired,
}
function BooleanOption({ value, onChange, ...props }) {
    return (
        <div className="checkbox">
            <label>
                <input type="checkbox"
                       checked={value}
                       onChange={e => onChange(e.target.checked)}
                       {...props}
                />
                Enable
            </label>
        </div>
    )
}

StringOption.propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
}
function StringOption({ value, onChange, ...props }) {
    return (
        <input type="text"
               value={value || ""}
               onChange={e => onChange(e.target.value)}
               {...props}
        />
    )
}
function Optional(Component) {
    return function ({ onChange, ...props }) {
        return <Component
            onChange={x => onChange(x ? x : null)}
            {...props}
        />
    }
}

NumberOption.propTypes = {
    value: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired,
}
function NumberOption({ value, onChange, ...props }) {
    return (
        <input type="number"
               value={value}
               onChange={(e) => onChange(parseInt(e.target.value))}
               {...props}
        />
    )
}

ChoicesOption.propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
}
export function ChoicesOption({ value, onChange, choices, ...props }) {
    return (
        <select
            onChange={(e) => onChange(e.target.value)}
            value={value}
            {...props}
        >
            { choices.map(
                choice => (
                    <option key={choice} value={choice}>{choice}</option>
                )
            )}
        </select>
    )
}

StringSequenceOption.propTypes = {
    value: PropTypes.arrayOf(PropTypes.string).isRequired,
    onChange: PropTypes.func.isRequired,
}
function StringSequenceOption({ value, onChange, ...props }) {
    const height = Math.max(value.length, 1)
    return <textarea
        rows={height}
        value={value.join('\n')}
        onChange={e => onChange(e.target.value.split("\n"))}
        {...props}
    />
}

export const Options = {
    "bool": BooleanOption,
    "str": StringOption,
    "int": NumberOption,
    "optional str": Optional(StringOption),
    "sequence of str": StringSequenceOption,
}

function PureOption({ choices, type, value, onChange, name, error }) {
    let Opt, props = {}
    if (choices) {
        Opt = ChoicesOption;
        props.choices = choices
    } else {
        Opt = Options[type]
    }
    if (Opt !== BooleanOption) {
        props.className = "form-control"
    }

    return <div className={classnames({'has-error':error})}>
                <Opt
                    name={name}
                    value={value}
                    onChange={onChange}
                    onKeyDown={stopPropagation}
                    {...props}
                />
            </div>
}
export default connect(
    (state, { name }) => ({
        ...state.options[name],
        ...state.ui.optionsEditor[name]
    }),
    (dispatch, { name }) => ({
        onChange: value => dispatch(updateOptions(name, value))
    })
)(PureOption)
