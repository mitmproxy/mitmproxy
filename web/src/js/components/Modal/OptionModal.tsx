/* eslint-disable react/prop-types */
import React from "react";
import { connect, shallowEqual } from "react-redux";
import * as modalAction from "../../ducks/ui/modal";
import { Option } from "../../ducks/options";
import { compact, isEmpty } from "lodash";
import { RootState, useAppDispatch, useAppSelector } from "../../ducks";
import OptionInput from "./OptionInput";

function OptionHelp({ name }: { name: Option }) {
    const help = useAppSelector((state) => state.options_meta[name]?.help);
    return <div className="help-block small">{help}</div>;
}

function OptionError({ name }) {
    const error = useAppSelector((state) => state.options_meta[name]?.error);
    if (!error) return null;
    return <div className="small text-danger">{error}</div>;
}

export function PureOptionDefault({ value, defaultVal }) {
    if (value === defaultVal) {
        return null;
    } else {
        if (typeof defaultVal === "boolean") {
            defaultVal = defaultVal ? "true" : "false";
        } else if (Array.isArray(defaultVal)) {
            if (
                isEmpty(compact(value)) && // filter the empty string in array
                isEmpty(defaultVal)
            ) {
                return null;
            }
            defaultVal = "[ ]";
        } else if (defaultVal === "") {
            defaultVal = '""';
        } else if (defaultVal === null) {
            defaultVal = "null";
        }
        return (
            <div className="small">
                Default: <strong> {defaultVal} </strong>{" "}
            </div>
        );
    }
}

const OptionDefault = connect(
    (state: RootState, { name }: { name: Option }) => ({
        value: state.options[name],
        defaultVal: state.options_meta[name]?.default,
    }),
)(PureOptionDefault);

export default function OptionModal() {
    const dispatch = useAppDispatch();
    const options = useAppSelector(
        (state) => Object.keys(state.options_meta),
        shallowEqual,
    ).sort() as Option[];

    return (
        <div>
            <div className="modal-header">
                <button
                    type="button"
                    className="close"
                    data-dismiss="modal"
                    onClick={() => dispatch(modalAction.hideModal())}
                >
                    <i className="fa fa-fw fa-times"></i>
                </button>
                <div className="modal-title">
                    <h4>Options</h4>
                </div>
            </div>

            <div className="modal-body">
                <div className="form-horizontal">
                    {options.map((name) => (
                        <div key={name} className="form-group">
                            <div className="col-xs-6">
                                <label htmlFor={name}>{name}</label>
                                <OptionHelp name={name} />
                            </div>
                            <div className="col-xs-6">
                                <OptionInput name={name} />
                                <OptionError name={name} />
                                <OptionDefault name={name} />
                            </div>
                        </div>
                    ))}
                </div>
            </div>

            <div className="modal-footer"></div>
        </div>
    );
}
