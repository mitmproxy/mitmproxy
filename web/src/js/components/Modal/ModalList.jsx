import React from 'react'
import ModalLayout from './ModalLayout'
import OptionContent from './OptionModal'
import CopyContent from './CopyModal'

function OptionModal() {
    return (
        <ModalLayout>
            <OptionContent/>
        </ModalLayout>
    )
}

function CopyModal() {
    return (
        <ModalLayout>
            <CopyContent/>
        </ModalLayout>
    )
}

export default [ OptionModal, CopyModal]
