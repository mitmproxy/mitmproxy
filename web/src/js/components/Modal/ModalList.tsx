import * as React from "react";
import ModalLayout from "./ModalLayout";
import OptionContent from "./OptionModal";

function OptionModal() {
    return (
        <ModalLayout>
            <OptionContent />
        </ModalLayout>
    );
}

export default [OptionModal];
