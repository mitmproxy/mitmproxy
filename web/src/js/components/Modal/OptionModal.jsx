import React, { Component } from "react"
import { connect } from "react-redux"
import * as modalAction from "../../ducks/ui/modal"
import * as optionAction from "../../ducks/options"
import Option from "./Option"
import _ from "lodash"

function PureOptionHelp({help}){
    return <div className="help-block small">{help}</div>;
}
const OptionHelp = connect((state, {name}) => ({
    help: state.options[name].help,
}))(PureOptionHelp);

function PureOptionError({error}){
    if(!error) return null;
    return <div className="small text-danger">{error}</div>;
}
const OptionError = connect((state, {name}) => ({
    error: state.ui.optionsEditor[name] && state.ui.optionsEditor[name].error
}))(PureOptionError);

export function PureOptionDefault({value, defaultVal}){
    if( value === defaultVal ) {
        return null
    } else {
        if (typeof(defaultVal) === 'boolean') {
            defaultVal = defaultVal ? 'true' : 'false'
        } else if (Array.isArray(defaultVal)){
            if (_.isEmpty(_.compact(value)) && // filter the empty string in array
                _.isEmpty(defaultVal)){
                return null
            }
            defaultVal = '[ ]'
        } else if (defaultVal === ''){
            defaultVal = '\"\"'
        } else if (defaultVal === null){
            defaultVal = 'null'
        }
        return <div className="small">Default: <strong> {defaultVal} </strong> </div>
    }
}
const OptionDefault = connect((state, {name}) => ({
    value: state.options[name].value,
    defaultVal: state.options[name].default
}))(PureOptionDefault)

class PureOptionModal extends Component {

    constructor(props, context) {
        super(props, context)
        this.state = { title: 'Options' }
    }

    componentWillUnmount(){
        // this.props.save()
    }

    render() {
        const { hideModal, options } = this.props
        const { title } = this.state
        return (
            <div>
                <div className="modal-header">
                    <button type="button" className="close" data-dismiss="modal" onClick={() => {
                        hideModal()
                    }}>
                        <i className="fa fa-fw fa-times"></i>
                    </button>
                    <div className="modal-title">
                        <h4>{ title }</h4>
                    </div>
                </div>

                <div className="modal-body">
                    <div className="form-horizontal">
                        {
                            options.map(name =>
                                <div key={name} className="form-group">
                                    <div className="col-xs-6">
                                        <label htmlFor={name}>{name}</label>
                                        <OptionHelp name={name}/>
                                    </div>
                                    <div className="col-xs-6">
                                        <Option name={name}/>
                                        <OptionError name={name}/>
                                        <OptionDefault name={name}/>
                                    </div>
                                </div>
                            )
                        }
                    </div>
                </div>

                <div className="modal-footer">
                </div>
            </div>
        )
    }
}

export default connect(
    state => ({
        options: Object.keys(state.options).sort()
    }),
    {
        hideModal: modalAction.hideModal,
        save: optionAction.save,
    }
)(PureOptionModal)
