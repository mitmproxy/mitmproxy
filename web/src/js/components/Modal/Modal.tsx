import * as React from "react"
import ModalList from './ModalList'
import { useAppSelector } from "../../ducks";


export default function PureModal() {
    const activeModal : string = useAppSelector(state => state.ui.modal.activeModal)
    const ActiveModal:(() => JSX.Element)  | undefined= ModalList.find(m => m.name === activeModal )

    return(
        activeModal&&ActiveModal!==undefined ? <ActiveModal/> : <div/>
    )
}
