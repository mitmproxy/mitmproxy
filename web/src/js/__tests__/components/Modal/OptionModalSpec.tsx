import * as React from "react";
import { PureOptionDefault } from "../../../components/Modal/OptionModal";
import { render } from "../../test-utils";

describe("PureOptionDefault Component", () => {
    it("should return null when the value is default", () => {
        const { asFragment } = render(
            <PureOptionDefault value="foo" defaultVal="foo" />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("should handle boolean type", () => {
        const { asFragment } = render(
            <PureOptionDefault value={true} defaultVal={false} />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("should handle array", () => {
        let a = [""],
            b = [],
            c = ["c"],
            { asFragment } = render(
                <PureOptionDefault value={a} defaultVal={b} />,
            );
        expect(asFragment()).toMatchSnapshot();

        asFragment = render(
            <PureOptionDefault value={a} defaultVal={c} />,
        ).asFragment;
        expect(asFragment()).toMatchSnapshot();
    });

    it("should handle string", () => {
        const { asFragment } = render(
            <PureOptionDefault value="foo" defaultVal="" />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("should handle null value", () => {
        const { asFragment } = render(
            <PureOptionDefault value="foo" defaultVal={null} />,
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
