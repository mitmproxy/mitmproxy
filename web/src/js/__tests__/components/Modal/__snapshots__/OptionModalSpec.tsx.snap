// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PureOptionDefault Component should handle array 1`] = `<DocumentFragment />`;

exports[`PureOptionDefault Component should handle array 2`] = `
<DocumentFragment>
  <div
    class="small"
  >
    Default: 
    <strong>
       [ ] 
    </strong>
     
  </div>
</DocumentFragment>
`;

exports[`PureOptionDefault Component should handle boolean type 1`] = `
<DocumentFragment>
  <div
    class="small"
  >
    Default: 
    <strong>
       false 
    </strong>
     
  </div>
</DocumentFragment>
`;

exports[`PureOptionDefault Component should handle null value 1`] = `
<DocumentFragment>
  <div
    class="small"
  >
    Default: 
    <strong>
       null 
    </strong>
     
  </div>
</DocumentFragment>
`;

exports[`PureOptionDefault Component should handle string 1`] = `
<DocumentFragment>
  <div
    class="small"
  >
    Default: 
    <strong>
       "" 
    </strong>
     
  </div>
</DocumentFragment>
`;

exports[`PureOptionDefault Component should return null when the value is default 1`] = `<DocumentFragment />`;
